### ISSUE TITLE
Reverse Proxy - allow overriding X-Forwarded-Host header

### ISSUE THREAD

**Issue by mkosieradzki** [Azure/service-fabric-issues#113]
Consider following configuration:
Internet Facing Proxy -> SF Reverse Proxy -> Stateful Service

If the Internet Facing Proxy sets X-Forwarded-Host header on SF Reverse Proxy Stateful Service receives two headers - in my opinion SF Reverse Proxy SHOULD NOT add X-Forwarded-Host header if it is provided by the caller.

Both current and proposed implementations might be also some kind of security issue if some implementation underneath trusts X-Forwarded-Host header from the SF Reverse Proxy, in a scenario where SF Reverse Proxy is internet facing (what IMO is a terrible idea).

Current implementation requires inventing new headers for standard situations (including #112 ).

**Comment by mkosieradzki** [comment]
Sorry for confusion. This header is propely set from the standard Host header provided to the Reverse Proxy and that's definitely proper behavior.

**Comment by mkosieradzki** [comment]
From my experiments:
1. X-Forwaded-Host is set based on Host header sent to proxy (and that's correct)
2. If X-Forwaded-Host header is added, there are two X-Forwarded-Host headers that might lead to potential security issue - there are two reasonable options:
a) completely ignoring X-Forwaded-Host - that would be consistent with the current behavior and would be secure if used as Internet Facing proxy
b) consistently Allow overriding so X-Forwarded-Host takes precedence over Host and that should be nice for multi-proxies (I believe most common) scenarios.
3. X-Forwaded-Proto is just ignored and set to protocol used for invoking SF Reverse Proxy. That is consistent with a but does not make any sense if SSL is terminated on the Internet Facing Proxy.

Because SF Reverse Proxy is actually internal SF cluster xplat communication facility not a full blown Internet Facing Proxy I would definitely prefer solution b:

So please allow overriding:
X-Forwarded-Host 
and 
X-Forwarded-Proto

**Comment by BharatNarasimman** [comment]
You shouldn't be seeing multiple x-forwarded-host headers. If there is an existing x-forwarded-host header, the value in HOST header is appended to that header's value. 
X-Forwarded-Proto is set to the protocol used for invoking the reverse proxy, so you woudnt know the protocol for all the proxies in the chain. 

The Forwarded header(https://tools.ietf.org/html/rfc7239) would be the most helpful, and that support is in the list of items in our backlog.

**Comment by mkosieradzki** [comment]
@BharatNarasimman I assumed that this appended value is actually from second header. But yes that's extactly what I am observing. I also have interpreted this as incorrect behavior, however I was wrong as the best reference for this non-RFC header I could find - http://httpd.apache.org/docs/2.2/mod/mod_proxy.html#x-headers states, that current SF proxy behavior is OK. 

> Be careful when using these headers on the origin server, since they will contain more than one (comma-separated) value if the original request already contained one of these headers

Forwarded-Host https://tools.ietf.org/html/rfc7239#section-5.3 is definitely more restrictive (it does not allow multiple host names)
>The syntax for a "host" value, after potential quoted-string unescaping, MUST conform to the Host ABNF described in Section 5.4 of [RFC7230].

Regarding X-Forwarded-Proto behavior I am not convinced that current behavior is reasonable. Even in the documentation you have pointed out Forwarded-Proto should contain outermost protocol for proper URL rewriting (https://tools.ietf.org/html/rfc7239#section-5.4 
> For example, in an environment where a reverse proxy is also used as a crypto offloader, this allows the origin server to rewrite URLs in a document to match the type of connection as the user agent requested, even though all connections to the origin server are unencrypted HTTP

To sum up, when implementing Forwarded-* headers please allow overriding for both headers as my original request stated so multi-level proxy configurations will work as expected. These headers are intended for URL rewriting so should contain public visible headers and SF proxy is not intended to be internet facing proxy by any means so situation with additional proxy layer on top of SF proxy will be the most common scenario.

**Comment by NArnott** [comment]
Any word on this? I'm currently stuck trying to run Identity Server 4 behind the SF reverse proxy due to this issue. 

We are using Nginx as our internet facing reverse proxy, which includes SSL termination. It sends the correct X-Forwarded-Proto as https, but the SF reverse proxy is overriding it to http. This causes the Identity Server's openid-configuration to show the endpoints as http, resulting clients to fail validation since the proto's don't match.

**Comment by MyCADDev** [comment]
Hi,
We are also stuck because of this issue. Is there any workaround?

**Comment by NArnott** [comment]
We ended up using our own custom headers to send this information from Nginx to our application, so that the SF Reverse Proxy wouldn't tamper with them.
